lib/snitch-core/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+        self.iter_contents()
+            .skip_while(move |rec| rec.ena < first_ena)
+    }
+
+    /// Discards records up through and including `last_written_ena`.
+    ///
+    /// This is intended to be called when we've received confirmation that a
+    /// range of records has been committed to a database. The provided ENA is
+    /// _inclusive_ and indicates the last record written to the database.
+    ///
+    /// If the ENA is either lower than any of our stored records, or higher
+    /// than what we've vended out, this is a no-op.
+    pub fn flush_thru(&mut self, last_written_ena: u64) {
+        let count = last_written_ena
+            .saturating_add(1)
+            .saturating_sub(self.earliest_ena);
+        for _ in 0..count {
+            // Discard the lead record in the queue.
+            let mut slices = self.storage.as_slices();
+            let len = u16::from_le_bytes(take_array(&mut slices).unwrap_lite())
+                as usize;
+
+            let size = OVERHEAD + len;
+            // So it's weird, but, heapless::Deque only lets you pop individual
+            // bytes. Hopefully this is not too expensive.
+            for _ in 0..size {
+                self.storage.pop_front();
+            }
+
+            self.earliest_ena += 1;
+        }
+    }
+
+    /// Makes a best effort at inserting a record.
+    ///
+    /// This will attempt to recover from the `Losing` state before writing
+    /// anything.
+    ///
+    /// If, after recovery, we are `Collecting` and there is enough space for a
+    /// header (`OVERHEAD`) plus `data`, the record is stored, implicitly
+    /// assigned the next ENA.
+    ///
+    /// If recovery is not possible, the existing loss count is advanced.
+    ///
+    /// If we are `Collecting` but `data` plus a header won't fit, we enter the
+    /// `Losing` state with a count of 1.
+    fn insert_impl(&mut self, sender: u16, timestamp: u64, data: &[u8]) {
+        let data_len = u16::try_from(data.len()).unwrap_lite();
+        self.recover_if_required();
+
+        match &mut self.insert_state {
+            InsertState::Collecting => {
+                let room = self.storage.capacity() - self.storage.len();
+                if room >= OVERHEAD + data.len() {
+                    self.write_header(sender, timestamp, data_len);
+                    for &byte in data {
+                        self.storage.push_back(byte).unwrap_lite();
+                    }
+                } else {
+                    self.insert_state = InsertState::Losing {
+                        count: NonZeroU32::new(1).unwrap_lite(),
+                        timestamp,
+                    };
+                }
+            }
+            InsertState::Losing { count, .. } => {
+                *count = count.saturating_add(1);
+            }
+        }
+    }
+
+    /// Internal utility routine for storing a record header.
+    fn write_header(&mut self, task: u16, timestamp: u64, data_len: u16) {
```

Comment:
Nit: is there a reason arguments here are in a different order than the header?

Reply:
Path dependence. ;-) This was actually the original order of fields in the header, until it became clear that len-first made the flush impl simpler. I'll swap 'em.